server/src/main/java/org/opensearch/search/query/BitmapIndexQuery.java

```diff
@@ -0,0 +1,266 @@
+            }
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                // This query depend only on segment-immutable structure points
+                return true;
+            }
+        };
+    }
+
+    private class MergePointVisitor implements PointValues.IntersectVisitor {
+        private final DocIdSetBuilder result;
+        private final BytesRefIterator iterator;
+        private BytesRef nextQueryPoint;
+        private final ArrayUtil.ByteArrayComparator comparator;
+        private DocIdSetBuilder.BulkAdder adder;
+
+        public MergePointVisitor(DocIdSetBuilder result)
+            throws IOException {
+            this.result = result;
+            this.comparator = ArrayUtil.getUnsignedComparator(Integer.BYTES);
+            this.iterator = bitmapEncodedIterator(bitmap);
+            nextQueryPoint = iterator.next();
+        }
+
+        @Override
+        public void grow(int count) {
+            adder = result.grow(count);
+        }
+
+        @Override
+        public void visit(int docID) {
+            adder.add(docID);
+        }
+
+        @Override
+        public void visit(DocIdSetIterator iterator) throws IOException {
+            adder.add(iterator);
+        }
+
+        @Override
+        public void visit(int docID, byte[] packedValue) {
+            if (matches(packedValue)) {
+                visit(docID);
+            }
+        }
+
+        @Override
+        public void visit(DocIdSetIterator iterator, byte[] packedValue) throws IOException {
+            if (matches(packedValue)) {
+                adder.add(iterator);
+            }
+        }
+
+        private boolean matches(byte[] packedValue) {
+            while (nextQueryPoint != null) {
+                int cmp = comparator.compare(nextQueryPoint.bytes, nextQueryPoint.offset, packedValue, 0);
+                if (cmp == 0) {
+                    return true;
+                } else if (cmp < 0) {
+                    // Query point is before index point, so we move to next query point
+                    try {
+                        nextQueryPoint = iterator.next();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                } else {
+                    // Query point is after index point, so we don't collect and we return:
+                    break;
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public PointValues.Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
+            while (nextQueryPoint != null) {
+                int cmpMin =
+                    comparator.compare(nextQueryPoint.bytes, nextQueryPoint.offset, minPackedValue, 0);
+                if (cmpMin < 0) {
+                    // query point is before the start of this cell
+                    try {
+                        nextQueryPoint = iterator.next();
```

Comment:
Can we skip straight to the first entry >= minPackedValue?

Reply:
Yes, now using the `PeekableIntIterator` from roaring bitmap that can advance.
Added a randomized unit test to make sure it's working correctly.